### ISSUE TITLE
TreeNode.lowest_common_ancestor only works with tips

### ISSUE THREAD

**Issue by wwood** [biocore/scikit-bio#807]
Hi,

So far I've found `skbio` quite helpful and stable, until I came across this. I hope this is a reasonably straightforward feature request, a small tweak of the current method.

Is there any reason why the `lca` method shouldn't work non-tips as well? I believe it currently fails when one node is the parent of another:
```
>>> from skbio.tree import TreeNode
>>> from StringIO import StringIO
>>> tree = TreeNode.read(StringIO("((A:1.0,B:2.0)'g__genus1':3.0)root;"))
>>> print tree.ascii_art()
                    /-A
-root---- /g__genus1
                    \-B
>>> tree.lca(['A','g__genus1'])
<TreeNode, name: root, internal node count: 1, tips count: 2>
```
(when the `g__genus1` node is expected)

Thanks in advance,
ben

**Comment by wwood** [comment]
Just realised also the `distance` method suffers from the same problem with that testcase, and there the documentation doesn't say `tips` only `nodes`, so unless I'm wrong this is now a bug report as well as a feature request. With the same tree as above,
```
>>> tree.find('A').distance(tree.find('g__genus1'))
7.0
```
(expected 1, 7 being 1+3+3)

**Comment by wasade** [comment]
w/o checking the code, I believe the LCA cannot be one of the nodes passed
in. I don't believe it is limited to working with just tips as it should
use `ancestors`

**Comment by wasade** [comment]
Odd about distance, will need to look at the code

**Comment by wwood** [comment]
Cannot seems almost always true, but _should_ not?

https://en.wikipedia.org/wiki/Lowest_common_ancestor
> the lowest common ancestor (LCA) of two nodes v and w in a tree or directed acyclic graph (DAG) is the lowest (i.e. deepest) node that has both v and w as descendants, where we define each node to be a descendant of itself (so if v has a direct connection from w, w is the lowest common ancestor).

If you stick to the definition that it shouldn't be one of the nodes passed in, then I've another bug for you - what if you pass in the root node?

`distance` calls `lca` so there's the problem with that.

**Comment by wasade** [comment]
Interesting, good catch :)

**Comment by jairideout** [comment]
Thanks for reporting this @wwood! Are you interested in contributing a pull request to fix this bug?

**Comment by ebolyen** [comment]
@wasade, @jairideout Would it be worth revisiting our implementation of TreeNode at this time? e.g. using NetworkX like we briefly discussed in Boulder a while back?

**Comment by jairideout** [comment]
:+1:, can you please create a separate issue for this?

**Comment by wasade** [comment]
:+1:

**Comment by ebolyen** [comment]
created #809

**Comment by jairideout** [comment]
Awesome, thanks!

**Comment by rob-knight** [comment]
Watch out for performance issues (although those may have been resolved now).


On Jan 20, 2015, at 9:00 PM, Evan Bolyen <notifications@github.com<mailto:notifications@github.com>> wrote:


@wasade<https://github.com/wasade>, @jairideout<https://github.com/jairideout> Would it be worth revisiting our implementation of TreeNode at this time? e.g. using NetworkX like we briefly discussed in Boulder a while back?

**Comment by ElDeveloper** [comment]
Indeed, Networkx has a few performance issues, not sure what
methods/functions you will be using but it's important to remember that
it's pure python so there may be things that could be faster by properly
using NumPy.

**Comment by ebolyen** [comment]
Currently TreeNode does not make significant use of NumPy, so NetworkX may in fact still be a faster pure python implementation than our pure python implementation. 

This is a good conversation, but let's direct it to #809

**Comment by jairideout** [comment]
@wasade are you available to work on these bugs for 0.5.0? We're shooting for June 7 to cut the release.

**Comment by wasade** [comment]
Would a sufficient resolution be to just disallow nontips? If so, then I can issue a PR on that in the next few days. Not able to invest much more though right now, sorry

**Comment by jairideout** [comment]
👍 for disallowing non-tips for now. `lca` and `distance` give incorrect results right now for this case, so having an error message would be much better than a silent bug. Thanks @wasade!

**Comment by wasade** [comment]
This is not actually an issue. However, it does seem to be the case that distance is incorrectly computed when the LCA of the nodes is one of the nodes to determine the LCA of. 

```python
16:30:36 (daniel@sandbar):~> t = skbio.TreeNode.read(io.StringIO('((a,b)c,(d,e)f)root;'))

16:40:49 (daniel@sandbar):~> print(t.ascii_art())
                    /-a
          /c-------|
         |          \-b
-root----|
         |          /-d
          \f-------|
                    \-e

16:40:53 (daniel@sandbar):~> t.lowest_common_ancestor(['f', 'a'])
Out[10]: <TreeNode, name: root, internal node count: 2, tips count: 4>
```

**Comment by jairideout** [comment]
I think the output is incorrect in @wwood's [original example](https://github.com/biocore/scikit-bio/issues/807#issue-54690349). It seems the bug happens in `lca` and `distance` when the LCA between two nodes is one of the nodes.

**Comment by wasade** [comment]
But, are you your own ancestor? If so, then the output is incorrect. If
not, then the output is correct. The issue with distance is related but
different.

**Comment by jairideout** [comment]
From one of @wwood's [previous comments](https://github.com/biocore/scikit-bio/issues/807#issuecomment-70403854), Wikipedia LCA seems to think you're your own ancestor. I don't know what the correct definition is but the docs should probably be updated at minimum if we decide this is a feature and not a bug.

**Comment by wasade** [comment]
Resolving this appropriately will represent a backwards incompatible change with respect to scikit-bio, and will make skbio inconsistent with PyCogent and ete3. Specifically, all three packages currently implement `ancestors` as not including `self`. So while the current implementation may not mesh with wikipedia, it is consistent with the other implementations. 

Note, I still do agree that the `distance` result is a bug regardless of how that may compute in cogent or ete3.

**Comment by jairideout** [comment]
👍 for matching LCA in PyCogent and ete3, thanks for checking!

**Comment by jairideout** [comment]
Fixed in #1367. Thanks for finding and reporting this @wwood!
